### DISCUSSION TITLE
Login via header value?

### DISCUSSION THREAD

**Discussion by icsy7867** [YOURLS/YOURLS#3020] [Plugins]
I have yourls running on a docker container, and I have another container acting as a proxy and provides SSO auth via SAML to the /admin page.  (This actually works quite well, and links work as intended, but have to SSO to get to the admin page).

As part of this system, I pass an authenticated user via a header variable.  I am trying to write a plugin where, if this value exists, we do all the login things.  However due to a lack of understanding of the login process, I could use some help.  I tried looking through the source code, and tried to recreate all the login-y things.

My plugin looks like this:
(I am quite sure I am missing a ton of things I actually need, just not sure where to start!)
```
<?php
$headers=getallheaders();
$user = $headers["X-WEBAUTH-USER"];

if ($user){

	if
		// Normal only: cookies
		( !yourls_is_API() &&
		  isset( $_COOKIE[ yourls_cookie_name() ] ) )
		{
			yourls_do_action( 'pre_login_cookie' );
			$unfiltered_valid = yourls_check_auth_cookie();
		}

	elseif
		// API or normal: login with username & pwd
		( isset( $_REQUEST['username'] ) && !empty( $_REQUEST['username'] )  ) )
		{
			yourls_do_action( 'pre_login_username_password' );
			$unfiltered_valid = true;
		}

	yourls_do_action( 'login' );

	// (Re)store encrypted cookie if needed
	if ( !yourls_is_API() ) {
		yourls_store_cookie( $user );

		// Login form : redirect to requested URL to avoid re-submitting the login form on page reload
		if( isset( $user ) && isset( $_SERVER['REQUEST_URI'] ) ) {
			yourls_redirect( yourls_sanitize_url_safe($_SERVER['REQUEST_URI']) );
		}
	}

}
```

**Comment by ozh** [comment]
I don't think you completely understood how plugins work.

You'll want to modify how function [`yourls_is_valid_user()`](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-auth.php#L24) works, so you will hook into one of the existing hooks in this function, depending how your plugin works (ie [`shunt_is_valid_user`](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-auth.php#L26), [`pre_login`](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-auth.php#L43), [`is_valid_user`](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-auth.php#L89), etc..)

A basic plugin to completely skip the normal auth process would be, for instance : 

```php
<?php
/*
Plugin Name: Custom Auth
Description: Auth differently
*/

yourls_add_filter( 'shunt_is_valid_user', 'my_custom_auth');

function my_custom_auth() {
    // do stuff to test the headers
    ...
    // if it's OK : return true : this will short circuit the authentication done by YOURLS
    // if it's not : return null : this will resume the YOURLS auth
    // see : https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-auth.php#L25-L29
}
```

There are [several plugins](https://github.com/YOURLS/awesome-yourls) dealing with different auth mechanism, I suggest you check how they're working and start from here

**Reply by icsy7867** [reply]
I totally don't.  I really appreciate the guidance.

**Reply by ozh** [reply]
Don't miss the beginner tutorial for plugins in the wiki
